### ISSUE TITLE
Unnecessary improper_ctypes warning for extern "Rust"

### ISSUE THREAD

**Issue by josephlr** [rust-lang/rust#64593]
The following code generates an `improper_ctypes` warning:
```rust
extern "Rust" {
    fn foo(dest: &mut [u8]) -> Result<(), Error>;
}
```
I would expect this sort of warning for an `extern "C"` declaration, but it seems incorrect for an `extern "Rust"` declaration. Am I misunderstanding something, or is this a warning that is safe to disable?

**Comment by dhardy** [comment]
For context, see: https://github.com/rust-random/getrandom/pull/109

`Error` is `pub struct Error(NonZeroU32);` so both this and `Result` are not C types.

**Comment by josephlr** [comment]
> `Error` is `pub struct Error(NonZeroU32);` so both this and `Result` are not C types.

Also the `&mut [u8]` is not a C type either. The main issue seems to be that the compiler checks all `extern` declarations as if they were `extern "C"`, which I don't _think_ is right (but `extern "Rust"` isn't documented as much as `extern "C"`).

**Comment by cuviper** [comment]
There are instances of this in `liballoc` and `libcore` too:

https://github.com/rust-lang/rust/blob/9ad1e7c46cf690b7ec6953b142430d21ca2d8799/src/liballoc/alloc.rs#L243-L247

https://github.com/rust-lang/rust/blob/9ad1e7c46cf690b7ec6953b142430d21ca2d8799/src/libcore/panicking.rs#L73-L78
